feat: Add the ability to communicate with Vault via mTLS

rodrigorfk opened the pull request:
<!--

Thanks for opening a pull request! Here are some tips to get everything merged smoothly:

1. Read our contributor guidelines: https://cert-manager.io/docs/contributing/

2. Make sure your commits are signed off: https://cert-manager.io/docs/contributing/sign-off/

3. If the PR is unfinished, raise it as a draft or prefix the title with "WIP:" so it's clear to everyone.

4. Be sure to allow edits from maintainers so it's easier for us to help: https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork

-->

### Pull Request Motivation

Vault server can be configured to strictly enforce clients to present client certificates while connecting to the server in the HTTPs transport layer.

It is possible to configure Vault to use a client certificate to secure the transport layer ([tcp listener documentation](https://www.vaultproject.io/docs/configuration/listener/tcp#tcp-listener-examples)):
```
listener "tcp" {
  tls_disable = false
  tls_cert_file = "/vault/tls/server.crt"
  tls_key_file = "/vault/tls/server.key"
  tls_client_ca_file = "/vault/tls/ca.crt"
  tls_require_and_verify_client_cert = true
}
```
When Vault is configured in the way above, there is no possibility to properly configure Vault Issuer/ClusterIssuer by using existing CRDs.

Fixes #6613

### Kind

feature

### Release Note

```release-note
Add the ability to communicate with Vault via mTLS when strict client certificates is enabled at Vault server side
```

inteon commented:
xref: https://github.com/cert-manager/cert-manager/pull/4330

wallrj commented:
/ok-to-test
/kind feature

wallrj commented on internal/apis/certmanager/types_issuer.go:
```diff
@@ -197,6 +197,16 @@ type VaultIssuer struct {
 	// +optional
 	CABundleSecretRef *cmmeta.SecretKeySelector
+
+	// Reference to a Secret containing a PEM-encoded Client Certificate to use when the
+	// Vault server requires mTLS.
+	// +optional
+	ClientCertSecretRef *cmmeta.SecretKeySelector
```
I **think** this should say "PEM-encoded Client Certificate **chain**" because the client needs to present the client certificate and any intermediate certificates that complete the chain to the CA certificate that the Vault server is using to verify the client cert.

> // Certificates contains one or more certificate chains to present to the
> // other side of the connection. The first certificate compatible with the
> // peer's requirements is selected automatically.

https://pkg.go.dev/crypto/tls#Config

wallrj replied:
I also wonder if if should say (and support) "chain**s**" (plural).
Consider the case where cert-manager is connecting to a  Vault **service** backed by a cluster of Vault servers.
If a new client CA certificate has been created and is being rolled out, the vault servers may be using a mixture of original and new client CA certificates, so the clients would each need client certificates chains for both the original and the new client CAs....at least until all the servers were using the new CA.
I don't know if that's a realistic scenario. Please explain briefly how you intend to rotate the client CA in your infrastructure.

rodrigorfk replied:
Hi @wallrj , thanks for your comment, while what you said that a chain of multiple client certificates could be provided, the client certificate don't have to present the CA certificate once again, client certificates only contains the certificate of the client itself, the CA certificate is configured at server side under the `tls_client_ca_file` configuration of the TCP listener as documented [here](https://developer.hashicorp.com/vault/docs/configuration/listener/tcp#tls_client_ca_file), and the server is responsible for validating if any of the client certificates provided matches the CA configured at server side.

Circling back to your suggestion of adding the word "chain" to the description, I can definitely do that, but realistic, based on what I said above, a client certificate is composed by only one certificate.

> Please explain briefly how you intend to rotate the client CA in your infrastructure.
- The initial work is done at Server side, you have to update the CA certificate bundle file used in the `tls_client_ca_file` configuration to include the new CA certificate chain as following:
  - you will keep the existing CA bundle in the file for existing client certificates to keep working
  - you will add the new CA bundle to get the server start trusting the new client certificates issued by it.
- Once you have the server updated with the new client CA bundle, you can start updating the client related secrets and adding the new client certificate and private keys, but once again, there is no need for providing the client CA bundled in the client certificate, the server already have it and will be responsible for validating the client certificate.

wallrj reviewed (changes requested):
Thanks @rodrigorfk 

This looks like a great start.

This feature will need documenting and that documentation will help me test this PR, so please open a cert-manager/website PR branched from the `release-next` branch, 
describing how to configure Vault for mTLS and how to configure the Vault issuer with a client certificate.
Explain how to quickly create a suitable client certificate,  using one of openssl, cfssl or step-cli.
Explain whether the client certificate is used for authentication , and if so which of the client certificate fields are used as the username for Vault.
Link to any relevant Vault documentation and tutorials on the subject.

You will need to rebase this PR, because I recently updated the vault E2E code to use vault-client-go rather than vault/api:
 * https://github.com/cert-manager/cert-manager/pull/6601

Find me in slack or come to one of our development meetings if you want to chat about any of this.
 * https://cert-manager.io/docs/contributing/#meetings
 * https://cert-manager.io/docs/contributing/#slack

rodrigorfk commented:
Thanks for your comments @wallrj , I will make sure I go over each one of them and proper address it.

rodrigorfk commented on test/e2e/framework/addon/vault/vault.go:
```diff
@@ -156,15 +171,24 @@ func (v *Vault) Setup(cfg *config.Config, leaderData ...internal.AddonTransferab
+					tls_require_and_verify_client_cert = %s
+				}`, strconv.FormatBool(v.EnforceMtls)),
+			},
+			{
+				Key:   "server.extraEnvironmentVars.VAULT_CLIENT_CERT",
+				Value: "/vault/tls/server.crt",
+			},
+			{
+				Key:   "server.extraEnvironmentVars.VAULT_CLIENT_KEY",
+				Value: "/vault/tls/server.key",
 			},
```
This perhaps looks weird, but it is essentially correct, that Vault server certificate was issued with the same CA which is now configured under the `tls_client_ca_file` configuration of the TCP listener, so they are completely valid client certificate as well, those environment variables are needed for the `exec` command in the Kubernetes Liveness and Readiness probe to work, the Probes are using the Vault CLI to check the status of the Vault server and therefore require a valid client certificate to interact with the status API at server side.

inteon replied:
Would it be possible you generate a new keypair instead of regenerating the existing pair?

rodrigorfk replied:
Sure, I am actually already generating a new certificate specifically for the clients in the e2e framework and storing it in a Kubernetes secret, I will then mount this secret in the Vault pod and use the client related certificate in those environment variables instead of the server certificate.

rodrigorfk replied:
@inteon and @wallrj, the configuration above was updated to use the client certificate in the environment variable, could you please have another look?

rodrigorfk commented on test/e2e/framework/addon/vault/setup.go:
```diff
@@ -208,7 +232,7 @@ func (v *VaultInitializer) Init() error {
 		}
 		var lastError error
-		err = wait.PollUntilContextTimeout(context.TODO(), time.Second, 20*time.Second, true, func(ctx context.Context) (bool, error) {
+		err = wait.PollUntilContextTimeout(context.TODO(), time.Second, 45*time.Second, true, func(ctx context.Context) (bool, error) {
```
This is to avoid a race condition while the Vault e2e addon is starting, [here](https://github.com/cert-manager/cert-manager/blob/b120669e8243b6ab868375495bbfc1e77d722be0/test/e2e/framework/addon/globals.go#L69) a new Vault addon with mTLS is synchronously started, however the creation of Vault client and the port-forward in the e2e framework was already dispatched asynchronously, the previous 20 seconds was a valid timeout when only one Vault addon was being created, since there is two now, that timeout has to be increased, otherwise the port-forward would fail because the second Vault addon was not yet ready.

inteon commented:
@rodrigorfk does this PR overlap with https://github.com/cert-manager/cert-manager/pull/4330?
If not, could you explain how it differs?

rodrigorfk commented:
> @rodrigorfk does this PR overlap with #4330? If not, could you explain how it differs?

Hi @inteon , thanks for your question, the implementation in both PRs could perhaps look similar, mostly because both cases are talking about client certificates, but they are essentially different cases, what is being implement in the #4330 so far is related to adding a new authentication method based on [TLS certificates auth method](https://developer.hashicorp.com/vault/docs/auth/cert), which is used to obtain the Vault Token from the `/v1/auth/cert/login` Vault API, what I'm proposing here is focusing solely in the transport layer and is not related to the Vault Token authentication, under certain strict compliance requirements, the Vault Server could be configure with [tls_require_and_verify_client_cert](https://developer.hashicorp.com/vault/docs/configuration/listener/tcp#tls_require_and_verify_client_cert) equals to `true`, so clients must present a client certificate when interacting will all Vault APIs, that does not perform any authentication to the APIs itself, it is just a complementary security mechanism at transport layer, the Vault Token must still be obtained from one one the supported authentication methods (Kubernetes, AppRole or Token). 

As I've mentioned in #6613, there was other issues, already closed, like #4774, #2926 and #4256 which were requesting a case similar to mine, once again, to related to the  `/v1/auth/cert/login` authentication method, but related to the whole transport layer, whereas client certificates must be present all the time.

inteon commented:
> > @rodrigorfk does this PR overlap with #4330? If not, could you explain how it differs?
> 
> Hi @inteon , thanks for your question, the implementation in both PRs could perhaps look similar, mostly because both cases are talking about client certificates, but they are essentially different cases, what is being implement in the #4330 so far is related to adding a new authentication method based on [TLS certificates auth method](https://developer.hashicorp.com/vault/docs/auth/cert), which is used to obtain the Vault Token from the `/v1/auth/cert/login` Vault API, what I'm proposing here is focusing solely in the transport layer and is not related to the Vault Token authentication, under certain strict compliance requirements, the Vault Server could be configure with [tls_require_and_verify_client_cert](https://developer.hashicorp.com/vault/docs/configuration/listener/tcp#tls_require_and_verify_client_cert) equals to `true`, so clients must present a client certificate when interacting will all Vault APIs, that does not perform any authentication to the APIs itself, it is just a complementary security mechanism at transport layer, the Vault Token must still be obtained from one one the supported authentication methods (Kubernetes, AppRole or Token).
> 
> As I've mentioned in #6613, there was other issues, already closed, like #4774, #2926 and #4256 which were requesting a case similar to mine, once again, ~to~ **not** related to the `/v1/auth/cert/login` authentication method, but related to the whole transport layer, whereas client certificates must be present all the time.

Thank you for the extra information.
It seems like the API looks different: `.clientCertSecretRef` and `.clientKeySecretRef` vs `.auth.clientCertificate`

rodrigorfk commented:
@wallrj the branch was rebased and a PR for the website repository was created [here](https://github.com/cert-manager/website/pull/1390) adding the documentation requested, I have addressed most of the comments and code changes from above, could you please have another look? thanks in advance.

rodrigorfk commented:
Hi @wallrj, thanks again for your initial review and comments, is there anything else you would like me to change or clarify? I would be willing to help on anyway possible to help you reviewing this change, FYI, I have been running internally in my organization a custom build of cert-manger v1.13.3 containing those changes and it is working pretty well.

inteon commented:
@rodrigorfk Great to hear you are actively testing this using real use cases. I think @wallrj said that he would review this PR.
I'll confirm that with him tomorrow.

wallrj reviewed (changes requested):
@rodrigorfk Sorry for delay.

I've finally got round to testing this and can confirm that it works for me with a Kind cluster and a Vault server manually installed and running in-cluster.

It seems that you have adjusted the E2E test framework to install two Vault servers:
the original vault server and a new one configured to only accept connections with client certificates.
And I think you have modified **one** set of conformance tests to target the mTLS server.
Have I understood correctly and is that what you intended?

I disagree with that change because it conflates the testing of two features.
And it is not clear from the test output that `Vault + client certificates` is being tested at all.

I suggest creating one new clearly documented E2E test for Vault + Client Cert and placing it in a new file in:
 * https://github.com/cert-manager/cert-manager/tree/master/test/e2e/suite/issuers/vault

rodrigorfk commented:
@wallrj thanks for circling back to this and for the thoughtful review.

> It seems that you have adjusted the E2E test framework to install two Vault servers:
the original vault server and a new one configured to only accept connections with client certificates.
And I think you have modified one set of conformance tests to target the mTLS server.
Have I understood correctly and is that what you intended?

yes, this statement is completely correct.

> I disagree with that change because it conflates the testing of two features.
And it is not clear from the test output that Vault + client certificates is being tested at all.
I suggest creating one new clearly documented E2E test for Vault + Client Cert

sure thing, the suggestion completely make sense, I was actually not sure if changing one existing conformance test would make it clear or not, I will follow your advise here and provide the test in a new file.

rodrigorfk commented:
@wallrj thanks once again for your review, I have addressed your suggestion related to move the mTLS related e2e test and created the `test/e2e/suite/issuers/vault/mtls.go` file containing it's cases. Could you please have another look?

wallrj commented on test/e2e/suite/issuers/vault/mtls.go:
Great. I can see the output of this new E2E test in the logs:
```
[cert-manager] Vault Issuer [mtls] should fail to init with missing Vault Token
/home/prow/go/src/github.com/cert-manager/cert-manager/test/e2e/suite/issuers/vault/mtls.go:165

  Timeline >>
  STEP: Creating a kubernetes client @ 02/15/24 17:35:25.635
  STEP: Creating an API extensions client @ 02/15/24 17:35:25.637
  STEP: Creating a cert manager client @ 02/15/24 17:35:25.637
  STEP: Creating a controller-runtime client @ 02/15/24 17:35:25.637
  STEP: Creating a gateway-api client @ 02/15/24 17:35:25.637
  STEP: Building a namespace api object @ 02/15/24 17:35:25.637
  STEP: Using the namespace e2e-tests-create-vault-issuer-99xmf @ 02/15/24 17:35:25.646
  STEP: Building a ResourceQuota api object @ 02/15/24 17:35:25.646
  STEP: Configuring the Vault server @ 02/15/24 17:35:25.664
  STEP: creating a service account for Vault authentication @ 02/15/24 17:35:29.948
  STEP: creating a client certificate for Vault mTLS @ 02/15/24 17:35:29.989
  STEP: Creating an Issuer @ 02/15/24 17:35:30.023
  STEP: Waiting for Issuer to become Ready @ 02/15/24 17:35:30.055
  Feb 15 17:35:30.055: INFO: Waiting for issuer test-vault-issuer condition v1.IssuerCondition{Type:"Ready", Status:"False", LastTransitionTime:<nil>, Reason:"", Message:"", ObservedGeneration:0}
  Feb 15 17:35:30.563: INFO: Waiting for issuer test-vault-issuer condition v1.IssuerCondition{Type:"Ready", Status:"False", LastTransitionTime:<nil>, Reason:"", Message:"", ObservedGeneration:0} (took 0s)
  STEP: Cleaning up AppRole @ 02/15/24 17:35:30.563
  STEP: Cleaning up Kubernetes @ 02/15/24 17:35:30.613
  STEP: Cleaning up Vault @ 02/15/24 17:35:30.633
  STEP: Deleting test namespace @ 02/15/24 17:35:30.661
  << Timeline
------------------------------
• [31.208 seconds]
[cert-manager] Vault Issuer [mtls] should be ready with a valid serviceAccountRef
/home/prow/go/src/github.com/cert-manager/cert-manager/test/e2e/suite/issuers/vault/mtls.go:440

  Timeline >>
  STEP: Creating a kubernetes client @ 02/15/24 17:34:59.497
  STEP: Creating an API extensions client @ 02/15/24 17:34:59.498
  STEP: Creating a cert manager client @ 02/15/24 17:34:59.498
  STEP: Creating a controller-runtime client @ 02/15/24 17:34:59.498
  STEP: Creating a gateway-api client @ 02/15/24 17:34:59.498
  STEP: Building a namespace api object @ 02/15/24 17:34:59.498
  STEP: Using the namespace e2e-tests-create-vault-issuer-hkxj2 @ 02/15/24 17:34:59.504
  STEP: Building a ResourceQuota api object @ 02/15/24 17:34:59.504
  STEP: Configuring the Vault server @ 02/15/24 17:34:59.519
  STEP: creating a service account for Vault authentication @ 02/15/24 17:35:29.842
  STEP: creating a client certificate for Vault mTLS @ 02/15/24 17:35:29.902
  STEP: Creating the Role and RoleBinding to let cert-manager use TokenRequest for the ServiceAccount @ 02/15/24 17:35:29.938
  STEP: Creating an Issuer @ 02/15/24 17:35:30.004
  STEP: Waiting for Issuer to become Ready @ 02/15/24 17:35:30.056
  Feb 15 17:35:30.056: INFO: Waiting for issuer test-vault-issuer condition v1.IssuerCondition{Type:"Ready", Status:"True", LastTransitionTime:<nil>, Reason:"", Message:"", ObservedGeneration:0}
  Feb 15 17:35:30.565: INFO: Waiting for issuer test-vault-issuer condition v1.IssuerCondition{Type:"Ready", Status:"True", LastTransitionTime:<nil>, Reason:"", Message:"", ObservedGeneration:0} (took 0s)
  STEP: Cleaning up AppRole @ 02/15/24 17:35:30.612
  STEP: Cleaning up Kubernetes @ 02/15/24 17:35:30.652
  STEP: Cleaning up Vault @ 02/15/24 17:35:30.657
  STEP: Deleting test namespace @ 02/15/24 17:35:30.691
  << Timeline
------------------------------
```
https://prow.build-infra.jetstack.net/view/gs/jetstack-logs/pr-logs/pull/cert-manager_cert-manager/6614/pull-cert-manager-master-e2e-v1-28/1758181625193041920

wallrj reviewed (approved):
Thanks @rodrigorfk 

I tested this again using the latest website documentation, 
and it all seems to work well.

Let's merge it and I'll ask you to do some more testing of this feature when we release the first cert-manager 1.15 alpha pre-release. 

And at that time, perhaps you'd help us to write  a nice release note for this feature too.

We appreciate your contribution.

/lgtm
